Create post feedback backend

mzaatar opened the pull request:
- Add new columns `SessionId` and DeviceId to `SessionFeedbackEntity`
- Add a new column DeviceId to `ConferenceFeedbackEntity`
- Add another constructor for `SessionFeedbackEntity` to accept the properties of the feedback request. It will insert null into `Time` column.
- Add another constructor for `ConferenceFeedbackEntity` to accept the properties of the feedback request. 
Note: keeping `Time` column to make backward compatible with the old Office forms way
- Create an Azure function `SubmitFeedback` that receives session feedback request from frontend and persists it:
    - in the `SessionFeedback` Azure Table if `IsConferenceFeedback` = false
    - in the `ConferenceFeedback` Azure Table if `IsConferenceFeedback` = true
- Get rid of warning (unused exception variable e)

mzaatar committed:
Add SubmitVote Azure function

mzaatar committed:
Add Azure parameters

mzaatar committed:
Merge branch 'master' of https://github.com/dddwa/ddd-backend into create-post-feedback-backend

mzaatar committed:
Some tweaks after yellow duck review

robdmoore commented on DDD.Core/AzureStorage/SessionFeedbackEntity.cs:
```diff
@@ -1,13 +1,29 @@
+            PartitionKey = conferenceInstance;
+            RowKey = Guid.NewGuid().ToString();
+            SessionId = sessionId;
+            Name = name;
```
This can probably be null too if there is a session id?

mzaatar replied:
This "name" variable is the attendee name who is giving the feedback.
Don't think we should remove it.

robdmoore commented on DDD.Functions.Extensions/KeyDatesConfig.cs:
```diff
@@ -50,6 +50,14 @@ public class KeyDatesConfig : Attribute
         public DateTimeOffset StopSyncingTitoFromDate => StopSyncingTitoFrom != null ? DateTimeOffset.Parse(StopSyncingTitoFrom) : DateTimeOffset.MinValue;
 
+        // Feedback
+        [AppSetting(Default = "FeedbackAvailableFrom")]
```
These variables need to be set for the test and prod sites via Azure DevOps release variables and passed through in the ARM and PowerShell

aidanmorgan replied:
Can we use the "on the day" setting we already have?

mzaatar replied:
I believe "on the day" is not available in the backend side. Unfortunately It is only front end :(

robdmoore commented on DDD.Functions/GetPrizeDraw.cs:
```diff
@@ -26,7 +26,8 @@ public static class GetPrizeDraw
             var sessionFeedback = await sessionFeedbackRepo.GetAllAsync(conference.ConferenceInstance);
-
+            
+            // we might need to use the SessionId instead of SessionName here
```
Yep, I think that's better than relying on name when we don't need to, just keep it backwards compatible.

mzaatar replied:
Kept as is but this comment maybe for future refactoring 👍

robdmoore commented on DDD.Functions/SubmitSessionFeedback.cs:
```diff
@@ -0,0 +1,91 @@
+    public class SessionFeedbackRequest
+    {
+        // not used for now
+        public Guid DeviceId { get; set; }
```
Add it to the entity so it's recorded. It'll be null for the microsoft forms way obv

mzaatar replied:
All good :) added

robdmoore commented on DDD.Functions/SubmitSessionFeedback.cs:
```diff
@@ -0,0 +1,91 @@
+        // not used for now
+        public Guid DeviceId { get; set; }
+        public string SessionId { get; set; }
+        public string Name { get; set; }
```
Get rid of name

mzaatar replied:
This is the name of the person giving the feedback :) How we could track them back?

robdmoore commented on DDD.Functions/SubmitSessionFeedback.cs:
```diff
@@ -0,0 +1,91 @@
+        public Guid DeviceId { get; set; }
+        public string SessionId { get; set; }
+        public string Name { get; set; }
+        public int Rating { get; set; }
```
Should we validate it's 1-5?

mzaatar replied:
I removed checking that rate should be between 1-5 based on my conversation with @aidanmorgan. We will allow it as number based on how we will design the front end side.

robdmoore commented on infrastructure/Deploy.ps1:
```diff
@@ -75,6 +75,8 @@ function Get-Parameters() {
     "stopSyncingAppInsightsFrom"        = $StopSyncingAppInsightsFrom;
     "stopSyncingAgendaFrom"             = $StopSyncingAgendaFrom;
     "sessionizeAgendaApiKey"            = $SessionizeAgendaApiKey;
+    "feedbackAvailableFrom"             = $FeedbackAvailableFrom;
```
Are you missing these being passed in as parameters? And also adding them to the local parameters template?

mzaatar replied:
They have been added to the parameters templates side but pending on this PR's approval so I can add them in DevOps

robdmoore replied:
I more meant this parameter needs to be passed in at the top of the file otherwise it's not defined?

Then if you add a parameter to deploy.ps1 you also need to add it to the local.parameters.template.json file

aidanmorgan commented on DDD.Functions/SubmitSessionFeedback.cs:
```diff
@@ -0,0 +1,91 @@
+            }
+
+            // Valid feedback 
+            if ((feedback.Rating < 1 || feedback.Rating > 5) && string.IsNullOrWhiteSpace(feedback.Likes) && string.IsNullOrWhiteSpace(feedback.ImprovementIdeas))
```
are we doing 1-5? I personally thing 1-5 is effectively meaningless, had good experience with -1,0,1 (so negative, neutral, positive), but happy to go with the consensus.

1-5 has issues with range compression/j curves (what really is the difference between a 5 star talk and a 4 star talk - ?) 1-5 also has issues with people needing to do more thinking, which leads to more dropoffs. 


Link if you're super bored :-)  https://pdfs.semanticscholar.org/9430/f3a76ce04cb9de4b74fd231e954c7386c96c.pdf?_ga=2.96423196.1646251308.1563496672-279858019.1563496672

mzaatar replied:
Removed the rating check so it can be any number :)

mzaatar committed:
persist the device id and remove rate 1-5 check

mzaatar committed:
Extend the function to be able to receive conference feedback as well based on a flag in the request and save it in the ConferenceFeedback Entity
